pkg/license/license.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package license
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/license"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// defaultOperatorLicenseLevel is the default license level when no operator license is installed
+	defaultOperatorLicenseLevel = "basic"
+	// licensingCfgMapName is the name of the config map used to store licensing information
+	licensingCfgMapName = "elastic-licensing"
+)
+
+// LicensingInfo represents information about the operator license including the total memory of all Elastic managed
+// components
+type LicensingInfo struct {
+	Timestamp               string `json:"timestamp"`
+	EckLicenseLevel         string `json:"eck_license_level"`
+	TotalManagedMemory      string `json:"total_managed_memory"`
+	EnterpriseResourceUnits string `json:"enterprise_resource_units"`
+}
+
+// LicensingResolver resolves the licensing information of the operator
+type LicensingResolver struct {
+	operatorNs string
+	client     k8s.Client
+}
+
+// ToInfo returns licensing information given the total memory of all Elastic managed components
+func (r LicensingResolver) ToInfo(totalMemory resource.Quantity) (LicensingInfo, error) {
+	eru := inEnterpriseResourceUnits(totalMemory)
+	memoryInGB := inGB(totalMemory)
+	licenseLevel, err := r.getOperatorLicenseLevel()
+	if err != nil {
+		return LicensingInfo{}, err
+	}
+
+	return LicensingInfo{
+		Timestamp:               time.Now().Format(time.RFC3339),
+		EckLicenseLevel:         licenseLevel,
+		TotalManagedMemory:      memoryInGB,
+		EnterpriseResourceUnits: eru,
+	}, nil
+}
+
+// Save updates or creates licensing information in a config map
+func (r LicensingResolver) Save(info LicensingInfo, operatorNs string) error {
+	data, err := info.toMap()
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("Saving", "license_info", info)
+	cm := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: operatorNs,
+			Name:      licensingCfgMapName,
+		},
+		Data: data,
+	}
+	err = r.client.Update(&cm)
```

Comment:
I guess we could technically check if an update is required to avoid the unnecessary call to the apiserver but I don't think that would be a big issue anyway.

Reply:
I think this version is more optimal since we will be updating the config map most of the time.